### PR TITLE
HBASE-23695 Fail gracefully if no category is present

### PR DESCRIPTION
**Pull request by joshelser**

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 32s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   5m 32s |  master passed  |
| +1 :green_heart: |  compile  |   0m 23s |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 28s |  master passed  |
| +1 :green_heart: |  shadedjars  |   4m 40s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 23s |  master passed  |
| +0 :ok: |  spotbugs  |   0m 51s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   0m 50s |  master passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   4m 51s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 24s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 24s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 26s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   4m 35s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |  15m 33s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| +1 :green_heart: |  javadoc  |   0m 22s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   0m 54s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   3m 18s |  hbase-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 15s |  The patch does not generate ASF License warnings.  |
|  |   |  49m 51s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1052/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1052 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 382dfa863487 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1052/out/precommit/personality/provided.sh |
| git revision | master / a44f3b50e4 |
| Default Java | 1.8.0_181 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1052/1/testReport/ |
| Max. process+thread count | 267 (vs. ulimit of 10000) |
| modules | C: hbase-common U: hbase-common |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1052/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

File: hbase-common/src/test/java/org/apache/hadoop/hbase/HBaseClassTestRuleChecker.java [hunk 1]
```diff
@@ -41,9 +41,12 @@
   @Override
   public void testStarted(Description description) throws Exception {
     Category[] categories = description.getTestClass().getAnnotationsByType(Category.class);
-    for (Class<?> c : categories[0].value()) {
-      if (c == IntegrationTests.class) {
-        return;
+    // Don't fail if there is a missing category
+    if (categories.length > 0) {
+      for (Class<?> c : categories[0].value()) {
```
Comment (busbey) [inline hbase-common/src/test/java/org/apache/hadoop/hbase/HBaseClassTestRuleChecker.java]: Shouldn't we check all the categories instead of the first?

**Comment by joshelser** [comment]
Thanks @busbey  and @virajjasani for the reviews as well! I didn't notice your approvals until after pushing (was going to push last night, but someone beat me to it and then I needed to rebase). Apologies for not getting your `Signed-off-by` in the commit message.

File: hbase-common/src/test/java/org/apache/hadoop/hbase/HBaseClassTestRuleChecker.java (hunk 1 above)
Comment (joshelser) [reply hbase-common/src/test/java/org/apache/hadoop/hbase/HBaseClassTestRuleChecker.java]: I'm not actually sure what all is allowed by `Category`. I would assume that this for-loop handles both `@Category(OneInterface.class)` and `@Category({InterfaceA.class,InterfaceB.class})`. I have to experiment to see what else is actually allowable.

**Comment by ndimiduk** [comment]
Weird. With or without this patch I get a `java.lang.ExceptionInInitializerError` with the cause being the line number of the test's `HBaseClassTestRule` `@ClassRule`.

**Comment by joshelser** [comment]
```
% mvn -X
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /usr/local/Cellar/maven/3.6.3/libexec
Java version: 1.8.0_211, vendor: Oracle Corporation, runtime: /Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.14.5", arch: "x86_64", family: "mac"
```

maybe some difference in built tools, @ndimiduk ? That is goofy that you see something else.

**Comment by bharathv** [comment]
@ndimiduk / @joshelser  Think there is another similar bug lurking around in HBaseClassTestRule. Tried it out of curiosity and here is the full stack trace. (see the underlying stack..)

```
java.lang.ExceptionInInitializerError
	at sun.misc.Unsafe.ensureClassInitialized(Native Method)
	at sun.reflect.UnsafeFieldAccessorFactory.newFieldAccessor(UnsafeFieldAccessorFactory.java:43)
	at sun.reflect.ReflectionFactory.newFieldAccessor(ReflectionFactory.java:156)
	at java.lang.reflect.Field.acquireFieldAccessor(Field.java:1088)
	at java.lang.reflect.Field.getFieldAccessor(Field.java:1069)
	at java.lang.reflect.Field.get(Field.java:393)
	at org.junit.runners.model.FrameworkField.get(FrameworkField.java:88)
	at org.junit.runners.model.TestClass.collectAnnotatedFieldValues(TestClass.java:249)
	at org.junit.runners.ParentRunner.classRules(ParentRunner.java:280)
	at org.junit.runners.ParentRunner.withClassRules(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.classBlock(ParentRunner.java:217)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:412)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 0
	at org.apache.hadoop.hbase.HBaseClassTestRule.getTimeoutInSeconds(HBaseClassTestRule.java:62)
	at org.apache.hadoop.hbase.HBaseClassTestRule.forClass(HBaseClassTestRule.java:77)
	at org.apache.hadoop.hbase.TestTableName.<clinit>(TestTableName.java:44)
	... 17 more

```

**Comment by joshelser** [comment]
Awesomes, @bharathv ! I'll spin out a new Jira issue and mention you all there :)

File: hbase-common/src/test/java/org/apache/hadoop/hbase/HBaseClassTestRuleChecker.java (hunk 1 above)
Comment (joshelser) [reply hbase-common/src/test/java/org/apache/hadoop/hbase/HBaseClassTestRuleChecker.java]: The `@Category` annotation is not `Repeatable` so we can only have a single one. Its value is `Class<?>[]`, so we can't provide anything else, e.g. an array of class<?>[] is disallowed. I think this is OK, but will add an explicit check on this :)

**Pull request closed without merging**